File: packages-gen/dart/libs/xt-ui/lib/ui/view/runtime.dart
import 'package:xtalk_lang/common-data.dart' as xtd;
import 'xtalk_ui/ui/view/polyfill.dart' as polyfill;
import 'xtalk_ui/ui/view/backend.dart' as backend;
import 'package:xtalk_substrate/view.dart' as view;





runtime_create(node, spec, render_fn, opts) {
  opts = (opts ?? <dynamic, dynamic>{});
  var registry = backend.registry(opts["overrides"],opts["polyfills"] ?? polyfill.registry());
  var local = <dynamic, dynamic>{};
  for(var entry_53183 in (spec["bindings"] ?? <dynamic, dynamic>{}).entries){
    var binding_id = entry_53183.key;
    var binding = entry_53183.value;
    if("local" == binding["source"]){
      local[binding_id] = binding["initial"];
    }
  };
  var runtime = <dynamic, dynamic>{
    "listener_id":opts["listener_id"] ?? ("wind/" + spec["id"] + "/" + (DateTime.now().millisecondsSinceEpoch).toString()),
    "render_fn":render_fn,
    "spec":spec,
    "space_id":opts["space_id"],
    "backend":"wind",
    "registry":registry,
    "subscription":null,
    "invalidate":null,
    "local":local,
    "local_revision":0,
    "node":node
  };
  runtime["dispatch"] = ((action_desc, event) {
    return view.dispatch(node,runtime["space_id"],action_desc,event,null);
  });
  return runtime;
}

snapshot(runtime) {
  var out = view.snapshot(runtime["node"],runtime["spec"]);
  xtd.obj_assign(out,runtime["local"]);
  return out;
}

local_set(runtime, binding_id, value) {
  runtime["local"][binding_id] = value;
  var revision = 1 + runtime["local_revision"];
  runtime["local_revision"] = revision;
  var invalidate = runtime["invalidate"];
  if((invalidate.runtimeType).toString().contains("Function") || (invalidate.runtimeType).toString().contains("=>") || (invalidate).toString().startsWith("Closure")){
    invalidate(
      snapshot(runtime),
      revision,
      <dynamic, dynamic>{"source":"local","binding_id":binding_id}
    );
  }
  return value;
}

resolve_node(runtime, node, seen) {
  if((null == node) || ("String" == (node.runtimeType).toString()) || (("int" == (node.runtimeType).toString()) || ("double" == (node.runtimeType).toString()) || ("num" == (node.runtimeType).toString()))){
    return node;
  }
  if((node.runtimeType).toString().startsWith("List") || (node.runtimeType).toString().startsWith("_GrowableList")){
    return xtd.arr_map(node,(child) {
      return resolve_node(runtime,child,<dynamic, dynamic>{});
    });
  }
  var component_id = node["component"];
  var registry = runtime["registry"];
  var override = (registry["overrides"])[component_id];
  var native = backend.native_entry(registry,component_id);
  if(((null != override) && (false != override)) || ((null != native) && (false != native))){
    return node;
  }
  if(seen.containsKey(component_id)){
    throw "view polyfill cycle [wind] - " + component_id;
  }
  var lowering = (registry["polyfills"])[component_id];
  if(!((lowering.runtimeType).toString().contains("Function") || (lowering.runtimeType).toString().contains("=>") || (lowering).toString().startsWith("Closure"))){
    throw "view implementation missing [wind] - " + component_id;
  }
  seen[component_id] = true;
  return resolve_node(runtime,lowering(node),seen);
}

prepare_node(runtime, node, state) {
  if(null == node){
    return null;
  }
  if(("String" == (node.runtimeType).toString()) || (("int" == (node.runtimeType).toString()) || ("double" == (node.runtimeType).toString()) || ("num" == (node.runtimeType).toString()))){
    return <dynamic, dynamic>{
      "type":"WText",
      "props":<dynamic, dynamic>{"text":(node).toString()},
      "children":<dynamic>[]
    };
  }
  if((node.runtimeType).toString().startsWith("List") || (node.runtimeType).toString().startsWith("_GrowableList")){
    return xtd.arr_map(node,(child) {
      return prepare_node(runtime,child,state);
    });
  }
  if(true == (node["props"] ?? <dynamic, dynamic>{})["hidden"]){
    return <dynamic, dynamic>{
      "type":"WText",
      "props":<dynamic, dynamic>{"text":""},
      "children":<dynamic>[]
    };
  }
  node = resolve_node(runtime,node,<dynamic, dynamic>{});
  var component_id = node["component"];
  var registry = runtime["registry"];
  var children = prepare_node(runtime,node["children"] ?? <dynamic>[],state);
  var override = (registry["overrides"])[component_id];
  if((override.runtimeType).toString().contains("Function") || (override.runtimeType).toString().contains("=>") || (override).toString().startsWith("Closure")){
    return override(runtime,node,children,state);
  }
  var entry = backend.native_entry(registry,component_id);
  return backend.prepare_native(
    runtime,
    component_id,
    entry,
    node["props"] ?? <dynamic, dynamic>{},
    children,
    state
  );
}

prepare(runtime) {
  var snapshot = snapshot(runtime);
  var concrete = runtime["render_fn"](snapshot);
  view.validate(view.view_spec(
    (runtime["spec"])["id"],
    (runtime["spec"])["bindings"],
    concrete
  ));
  var state = <dynamic, dynamic>{"next":0,"actions":<dynamic, dynamic>{}};
  var json = prepare_node(runtime,concrete,state);
  return <dynamic, dynamic>{"json":json,"actions":state["actions"]};
}

open(runtime, notify) {
  var spec = runtime["spec"];
  runtime["invalidate"] = notify;
  var subscription = view.subscribe(runtime["node"],spec,runtime["listener_id"],(snapshot, revision, event) {
    return notify(snapshot,revision,event);
  });
  runtime["subscription"] = subscription;
  return runtime;
}

close(runtime) {
  var subscription = runtime["subscription"];
  if((null != subscription) && (false != subscription)){
    view.unsubscribe(subscription);
    runtime["subscription"] = null;
  }
  runtime["invalidate"] = null;
  return true;
}

File: packages-gen/js/libs/xt-ui/react/view/runtime.js
const React = require("react")

const backend = require("@xtalk/ui/react/view/backend.js")

const r = require("@xtalk/ui/react.js")

const polyfill = require("@xtalk/ui/react/view/polyfill.js")

const view = require("@xtalk/substrate/view.js")

function runtime_create(node,spec,render_fn,opts){
  opts = (opts || {});
  let registry = backend.registry(opts["overrides"],opts["polyfills"] || polyfill.registry());
  let local = {};
  for(let [binding_id,binding] of Object.entries(spec["bindings"] || {})){
    if("local" == binding["source"]){
      local[binding_id] = binding["initial"];
    }
  };
  let runtime = {
    "listener_id":opts["listener_id"] || ("react/" + spec["id"] + "/" + String(Date.now())),
    "render_fn":render_fn,
    "spec":spec,
    "space_id":opts["space_id"],
    "backend":"react",
    "registry":registry,
    "invalidate":null,
    "local":local,
    "node":node
  };
  runtime["dispatch"] = (function (action_desc,event){
    return view.dispatch(node,runtime["space_id"],action_desc,event,null);
  });
  return runtime;
}

function snapshot(runtime){
  let out = view.snapshot(runtime["node"],runtime["spec"]);
  Object.assign(out,runtime["local"]);
  return out;
}

function local_set(runtime,binding_id,value){
  runtime["local"][binding_id] = value;
  let invalidate = runtime["invalidate"];
  if("function" == (typeof invalidate)){
    invalidate();
  }
  return value;
}

function resolve_node(runtime,node,seen){
  if((null == node) || ("string" == (typeof node)) || ("number" == (typeof node))){
    return node;
  }
  if(Array.isArray(node)){
    return node.map(function (child){
      return resolve_node(runtime,child,{});
    });
  }
  let component_id = node["component"];
  let registry = runtime["registry"];
  let override = (registry["overrides"])[component_id];
  let native = backend.native_entry(registry,component_id);
  if(override || native){
    return node;
  }
  if(null != seen[component_id]){
    throw "view polyfill cycle [react] - " + component_id;
  }
  let lowering = (registry["polyfills"])[component_id];
  if(!("function" == (typeof lowering))){
    throw "view implementation missing [react] - " + component_id;
  }
  seen[component_id] = true;
  return resolve_node(runtime,lowering(node),seen);
}

function render_node(runtime,node){
  if(null == node){
    return null;
  }
  if("string" == (typeof node)){
    return node;
  }
  if("number" == (typeof node)){
    return String(node);
  }
  if(Array.isArray(node)){
    return node.map(function (child){
      return render_node(runtime,child);
    });
  }
  if(true == (node["props"] || {})["hidden"]){
    return null;
  }
  node = resolve_node(runtime,node,{});
  let component_id = node["component"];
  let registry = runtime["registry"];
  let props = node["props"] || {};
  let children = render_node(runtime,node["children"] || []);
  let override = (registry["overrides"])[component_id];
  if("function" == (typeof override)){
    return override(runtime,props,children);
  }
  let entry = backend.native_entry(registry,component_id);
  return backend.render_native(runtime,entry,props,children);
}

function render(runtime,snapshot){
  let concrete = runtime["render_fn"](snapshot);
  view.validate(view.view_spec(
    (runtime["spec"])["id"],
    (runtime["spec"])["bindings"],
    concrete
  ));
  return render_node(runtime,concrete);
}

function View({node,options,render_fn,spec}){
  let refresh = r.useRefresh();
  let runtime_ref = React.useRef(runtime_create(node,spec,render_fn,options));
  let subscription_ref = React.useRef(null);
  runtime_ref.current["invalidate"] = refresh;
  React.useEffect(function (){
    let subscription = view.subscribe(node,spec,(runtime_ref.current)["listener_id"],function (_snapshot,_revision,_event){
      return refresh();
    });
    subscription_ref.current = subscription;
    return function (){
      let current = subscription_ref.current;
      if(current){
        view.unsubscribe(current);
      }
    };
  },[]);
  let runtime = runtime_ref.current;
  return render(runtime,snapshot(runtime));
}

module.exports = {
  ["runtime_create"]:runtime_create,
  ["snapshot"]:snapshot,
  ["local_set"]:local_set,
  ["resolve_node"]:resolve_node,
  ["render_node"]:render_node,
  ["render"]:render,
  ["View"]:View
}